tests/__init__.py

```diff
@@ -320,3 +328,102 @@ def stream(self, **kwargs):
         while contents:
             yield contents
             contents = self.read()
+
+
+class FakeApplicationInput(prompt_toolkit.input.DummyInput):
+    def fileno(self):
+        return 0
+
+    @property
+    def closed(self):
+        return False
+
+
+@dataclasses.dataclass
+class AppKeyPressAction:
+    key: str
+
+
+@dataclasses.dataclass
+class AppAssertionAction:
+    assertion: Callable[[prompt_toolkit.Application], None]
+    failure_message_format: Optional[str] = None
+
+
+@dataclasses.dataclass
+class AppCallbackAction:
+    callback: Callable[[prompt_toolkit.Application], None]
+
+
+class PromptToolkitApplicationStubber:
+    def __init__(self, app):
+        self._app = app
+        self._app.input = FakeApplicationInput()
+        self._queue = []
+
+    def add_keypress(self, key, app_assertion=None):
+        self._queue.append(AppKeyPressAction(key))
+        if app_assertion:
+            failure_message_format = (
+                f'Incorrect action on key press "{key}": '
+                '{message}'
```

Comment:
Need to add `f` on this iine as well

Reply:
Oh that omission was on purpose. Mainly the `{message}` is filled out later when the `AssertionError` is propagated.